algo/src/main/java/org/neo4j/graphalgo/impl/BetweennessCentrality.java

```diff
@@ -0,0 +1,151 @@
+    private final Path[] paths;
+
+    public BetweennessCentrality(Graph graph) {
+        this.graph = graph;
+        this.centrality = new double[graph.nodeCount()];
+        this.stack = new IntStack();
+        this.sigma = new int[graph.nodeCount()];
+        this.d = new int[graph.nodeCount()];
+        queue = new IntArrayDeque();
+        paths = new Path[graph.nodeCount()];
+        delta = new double[graph.nodeCount()];
+    }
+
+    /**
+     * compute centrality
+     * @return itself for method chaining
+     */
+    public BetweennessCentrality compute() {
+        graph.forEachNode(this::compute);
+        return this;
+    }
+
+    /**
+     * iterate over each result until every node has
+     * been visited or the consumer returns false
+     *
+     * @param consumer the result consumer
+     */
+    public void forEach(ResultConsumer consumer) {
+        for (int i = graph.nodeCount() - 1; i >= 0; i--) {
+            if (!consumer.consume(graph.toOriginalNodeId(i), centrality[i])) {
+                return;
+            }
+        }
+    }
+
+    public Stream<Result> resultStream() {
+        return IntStream.range(0, graph.nodeCount())
+                .mapToObj(nodeId ->
+                        new Result(
+                                graph.toOriginalNodeId(nodeId),
+                                centrality[nodeId]));
+    }
+
+    private void compute(int startNode) {
+        clearPaths();
+        stack.clear();
+        queue.clear();
+        Arrays.fill(sigma, 0);
+        Arrays.fill(delta, 0);
+        Arrays.fill(d, -1);
+        sigma[startNode] = 1;
+        d[startNode] = 0;
+        queue.addLast(startNode);
+        while (!queue.isEmpty()) {
+            int node = queue.removeLast();
+            stack.push(node);
+            graph.forEachRelationship(node, Direction.OUTGOING, (source, target, relationId) -> {
+                if (d[target] < 0) {
+                    queue.addLast(target);
+                    d[target] = d[node] + 1;
+                }
+                if (d[target] == d[node] + 1) {
+                    sigma[target] += sigma[node];
+                    append(target, node);
+                }
+                return true;
+            });
+        }
+        while (!stack.isEmpty()) {
+            final int node = stack.pop();
+            if (null == paths[node]) {
+                continue;
+            }
+            paths[node].forEch(v -> {
+                delta[v] += (double) sigma[v] / (double) sigma[node] * (delta[node] + 1.0);
+                if (node != startNode) {
+                    centrality[node] += delta[node];
+                }
+                return true;
+            });
+        }
+    }
+
+    /**
+     * append nodeId to path
+     *
+     * @param path the selected path
+     * @param nodeId the node id
+     */
+    private void append(int path, int nodeId) {
+        if (null == paths[path]) {
+            paths[path] = new Path();
```

Comment:
This is just my first poc. I'd like to implement paths on a preallocated array if possible. but lets have a talk with @knutwalker before